### PR TITLE
meta: add 'desktop' entry for apps

### PR DESCRIPTION
**Pull request by oSoMoN**
LP: #1588359

### CODE REVIEW

File: snapcraft/internal/meta.py [hunk 1]
```diff
@@ -122,10 +123,15 @@ def _setup_from_setup(self):
...
         gui_src = os.path.join(setup_dir, 'gui')
         gui_dst = os.path.join(self.meta_dir, 'gui')
-        if os.path.exists(gui_dst) and os.path.exists(gui_src):
-            shutil.rmtree(gui_dst)
         if os.path.exists(gui_src):
-            shutil.copytree(gui_src, gui_dst)
+            for f in os.listdir(gui_src):
+                if os.path.splitext(f)[1] == '.desktop':
+                    logger.warning(
+                        "DEPRECATED: use the 'desktop' key for apps instead "
```
Comment (sergiusens) [inline snapcraft/internal/meta.py]: I wouldn't necessarily deprecate this, new projects might not care about existing conventions.

File: schema/snapcraft.yaml [hunk 2]
```diff
@@ -88,6 +88,12 @@ properties:
           stop-command: ["daemon"]
         additionalProperties: false
         properties:
+          desktop:
+            type: string
+            description: path to a desktop file representing the app, relative to the prime directory
+          icon:
```
Comment (sergiusens) [inline schema/snapcraft.yaml]: is this here to support replacing icon names in existing desktop files instead of paths?

File: snapcraft/internal/meta.py (hunk 1 above)
Comment (oSoMoN) [inline snapcraft/internal/meta.py]: I don’t have a strong opinion on this, so I can undeprecate it if you wish.

File: schema/snapcraft.yaml (hunk 2 above)
Comment (oSoMoN) [inline schema/snapcraft.yaml]: This is here to allow overriding the icon in the upstream desktop file with a custom one for the snap.

For example, for webbrowser-app the Icon line for the input desktop file (before transformation) is:

    Icon=@CMAKE_INSTALL_FULL_DATADIR@/webbrowser-app/webbrowser-app.png

Typically on a deb system that would end up being:

    Icon=/usr/share/webbrowser-app/webbrowser-app.png

But when building a snap from the source tree, the line in the resulting desktop file in the prime directory will look like this:

    Icon=/share/webbrowser-app/webbrowser-app.png

Allowing to override the icon in snapcraft.yaml means upstream doesn’t need to care to support snaps.

Comment (sergiusens) [inline schema/snapcraft.yaml]: That last case is where I'd like to be smart, detect it is a path and apply the correct snapcraft magic foo of prefixing $SNAP in that path.

File: snapcraft/internal/meta.py [hunk 3]
```diff
@@ -215,19 +217,62 @@ def _wrap_exe(self, command, basename=None):
...
+            try:
+                app[k] = self._wrap_exe(app[k], '{}-{}'.format(k, name))
+            except CommandError as e:
+                raise EnvironmentError(
+                    'The specified command {!r} defined in the app {!r} '
+                    'does not exist or is not executable'.format(str(e), name))
+        if 'desktop' in app:
+            self._wrap_desktop_and_icon(name, app)
+
+    def _wrap_desktop_and_icon(self, name, app):
```
Comment (sergiusens) [inline snapcraft/internal/meta.py]: we actually reformat the desktop file instead of wrapping it. If you want to change the name to reflect that it would be great

File: snapcraft/internal/meta.py [hunk 4]
```diff
@@ -215,19 +217,62 @@ def _wrap_exe(self, command, basename=None):
...
+                'desktop file'.format(desktop_file))
+        if 'Exec' not in desktop_contents[section]:
+            raise EnvironmentError(
+                'The specified desktop file {!r} is missing the '
+                '"Exec" key'.format(desktop_file))
+        # XXX: do we want to allow more parameters for Exec?
+        exec_value = '{}.{} %U'.format(self._config_data['name'], name)
+        desktop_contents[section]['Exec'] = exec_value
+        if 'Icon' in desktop_contents[section]:
+            icon = desktop_contents[section]['Icon']
```
Comment (sergiusens) [inline snapcraft/internal/meta.py]: would be good to fail or warn if it does not exist

Comment (sergiusens) [inline snapcraft/internal/meta.py]: maybe fallback to the snap icon in meta/gui/icon if it exists, but this would be a different PR ;-)

Comment (oSoMoN) [inline snapcraft/internal/meta.py]: Done, a warning will now be issued if the icon does not exist.

Comment (oSoMoN) [inline snapcraft/internal/meta.py]: And I can confirm that I tested this branch on the webbrowser-app snap, and it works as expected. I haven’t tested on other snaps though.
